### PR TITLE
feat: adding option to force incremental cumulative propagators to find the same trace as non-incremental

### PR DESCRIPTION
**Pull request by ImkoMarijnissen**
It is difficult to compare the different types of changes made to the cumulative due to the fact that traces oftentimes diverge due to differences in the explanation(s)

This PR addresses this issue by creating a synchronisation between the incremental and the non-incremental version of the time-tabling propagators by ensuring that the explanations for the conflicts and the propagations are the same as would be found by the non-incremental propagator and the incremental version of the propagator.

For both the per-point reasoning and the over-interval reasoning, a `synchronisation` module is created which finds the correct conflicts and orders the profile tasks in the profile in the correct order!

_Edit: After discussion with @maartenflippo, this PR now also contains a new crate which contains procedural macros; these are currently used to create test cases for the cumulative where we want to test a lot of combinations of tests!_

### CODE REVIEW

File: pumpkin-solver/src/bin/pumpkin-solver/main.rs [hunk 1]
```diff
@@ -342,7 +342,9 @@ struct Args {
     /// Currently, the solver only supports variations on time-tabling methods.
     ///
     /// Possible values: ["time-table-per-point", "time-table-per-point-incremental",
-    /// "time-table-over-interval", "time-table-over-interval-incremental"]
+    /// "time-table-per-point-incremental-synchronised",
+    /// "time-table-over-interval", "time-table-over-interval-incremental",
+    /// "time-table-over-interval-incremental-synchronised"]
     #[arg(long = "cumulative-propagation-method", value_parser = cumulative_propagation_method_parser, default_value_t = CumulativePropagationMethod::default())]
```
Comment (maartenflippo) [inline pumpkin-solver/src/bin/pumpkin-solver/main.rs:348]: Use `ValueEnum`?

File: pumpkin-solver/src/propagators/arithmetic/absolute_value.rs [hunk 2]
```diff
@@ -2,10 +2,10 @@ use crate::basic_types::PropagationStatusCP;
 use crate::conjunction;
 use crate::engine::cp::propagation::ReadDomains;
 use crate::engine::domain_events::DomainEvents;
+use crate::engine::propagation::propagator_initialisation_context::PropagatorInitialisationContext;
```
Comment (maartenflippo) [inline pumpkin-solver/src/propagators/arithmetic/absolute_value.rs:5]: Plz remove changes

**Review by maartenflippo — CHANGES_REQUESTED** [review]

File: pumpkin-solver/tests/mzn_constraints/cumulative.template [hunk 3]
```diff
@@ -1,14 +0,0 @@
-predicate gecode_cumulatives(
```
Comment (ImkoMarijnissen) [inline pumpkin-solver/tests/mzn_constraints/cumulative.template:1]: @maartenflippo: add back this file

File: pumpkin-solver/Cargo.toml [hunk 4]
```diff
@@ -23,6 +23,7 @@ itertools = "0.13.0"
 flatzinc = "0.3.21"
 clap = { version = "4.5.17", features = ["derive"] }
 env_logger = "0.10.0"
+pumpkin-macros = { version = "0.1.0", path = "../pumpkin-macros"}
```
Comment (maartenflippo) [inline pumpkin-solver/Cargo.toml:26]: This can be a dev dependency

**Review by maartenflippo — DISMISSED** [review]

**Review by maartenflippo — APPROVED** [review]

**Pull request merged**